KAFKA-7883 add schema.namespace support to SetSchemaMetadata SMT in Kafka Connect

mnegodaev opened the pull request:
When you use Kafka Connect and Avro Converter with Schema Registry you may need to transform your source schema namespace. Currently `SetSchemaMetadata` provide a way to transform the namespace by specifying in the `schema.name` attribute both actual namespace being concatenated with the actual schema name:
```javascript
{
  "transforms" : "TransformSchema",
  "transforms.TransformSchema.type" : "org.apache.kafka.connect.transforms.SetSchemaMetadata$Value",
  "transforms.TransformSchema.schema.name" : "my.new.namespace.NewSchemaName"
}
```
But this way you are forced to specify the schema name even when you want to preserve the original schema name. In some cases you just may need to replace possibly an empty namespace of your source with the actual namespace fitting "your.java.package" and leave schema name as it is.
Moreover as mentioned in [KAFKA-7883](https://issues.apache.org/jira/browse/KAFKA-7883), when use JDBC connector with multiple tables in the source database what you may want is to preserve the table names but specify some non-empty namespace for them.
To support this we have to provide a way to append a schema namespace to the schema name (doesn't matter being updated or not).
In this PR a new `schema.namespace` attribute is added to `SetSchemaMetadata` letting developer to decide either to transform namespace only or transform both schema name and schema namespace by using either `schema.namespace` together with `schema.name` or just single `schema.name` attribute. An example:
```javascript
{
  "transforms" : "TransformSchema",
  "transforms.TransformSchema.type" : "org.apache.kafka.connect.transforms.SetSchemaMetadata$Value",
  "transforms.TransformSchema.schema.namespace" : "my.new.namespace",
  "transforms.TransformSchema.schema.name" : "NewSchemaName",
  "transforms.TransformSchema.schema.version" : 42
}
```

Commit:
KAFKA-7883 add schema.namespace support to SetSchemaMetadata SMT in Kafka Connect

Commit:
KAFKA-7883 test method refactoring

change test method signature to accept schema instead of Struct as it works with schemas only

Commit:
Merge remote-tracking branch 'upstream/trunk' into feature/KAFKA-7883

Commit:
Merge remote-tracking branch 'upstream/trunk' into feature/KAFKA-7883

Commit:
Merge remote-tracking branch 'upstream/trunk' into feature/KAFKA-7883

pjmagee commented:
I really want this! This is exactly what our team needs to resolve issues with Avro namespaces.

mnegodaev commented:
@kkonstantine could you please find a time to review my request? I believe it can be a useful feature and it should fill in the gaps in the current SetSchemaMetadata SMT API. Thank you!

pjmagee commented:
For context, we have had to go from 2 connectors to a.... huge number of connectors as a workaround, which has been a crazy amount of configuration for us, based on how we currently use kafka connect with yaml configuration (think of strimzi or the kafka confluent k8s operator). All each overriding the entire Schema name and being explicit while also introducing tabe whitelisting.... This pull request is a fantastic solution that would allow us to go back to 2 connectors.

rhauch commented:
I've marked this as needing a KIP, though I'm not yet sure whether that is the case. These SMTs are technically part of the API, since they appear in STM configurations, which are considered part of the public API. I'll keep looking into this to figure out if this is really the case.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/SetSchemaMetadata.java:
```diff
@@ -35,29 +37,34 @@
 
     public static final ConfigDef CONFIG_DEF = new ConfigDef()
             .define(ConfigName.SCHEMA_NAME, ConfigDef.Type.STRING, null, ConfigDef.Importance.HIGH, "Schema name to set.")
+            .define(ConfigName.SCHEMA_NAMESPACE, ConfigDef.Type.STRING, null, ConfigDef.Importance.HIGH, "Namespace to append to the schema name.")
```
Nit: aren't we _prepending_ the namespace, rather than _appending_ the namespace?
```suggestion
            .define(ConfigName.SCHEMA_NAMESPACE, ConfigDef.Type.STRING, null, ConfigDef.Importance.HIGH, "Namespace to prepend to the schema name.")
```

pjmagee replied:
Or just setting the entire value. Not prepending or appending a value. Would we not be setting a non-existing "namespace" field in the avro schema? With our jdbc connector, it doesnt create the namespace field at all. Its only if the schema name contains a '.' that it produces one.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/SetSchemaMetadata.java:
```diff
@@ -154,4 +161,14 @@ protected static Object updateSchemaIn(Object keyOrValue, Schema updatedSchema)
+        if (Utils.isBlank(namespace)) {
+            return name;
+        } else if (!Utils.isBlank(name)) {
+            return namespace + "." + name;
```
What happens if the `namespace` ends with a `.`? We should probably handle that case more transparently. Unit tests should verify that `schema.namespace=a.b.c.` and `schema.namespace=a.b.c` are treated the same.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/SetSchemaMetadata.java:
```diff
@@ -154,4 +161,14 @@ protected static Object updateSchemaIn(Object keyOrValue, Schema updatedSchema)
         return keyOrValue;
     }
+
+    private static String buildSchemaName(String namespace, String name) {
```
I know that we're trying to cover all of the cases in the tests that call the transform, but it'd be good to also have test case methods that check the different permutations of inputs to this method, including null, empty strings, and blank strings for names and namespaces, namespaces with and without a trailing `.`, namespaces with multiple `.` characters, namespaces with leading and/or trailing whitespaces, etc. 

In addition to such tests being easier to write and understand, they help test all possible combinations or code paths _and_ they help identify expected behaviors in various scenarios to better catch regressions in the future.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/SetSchemaMetadata.java:
```diff
@@ -71,7 +78,7 @@ public R apply(R record) {
                 schema.isOptional(),
                 schema.defaultValue(),
-                schemaName != null ? schemaName : schema.name(),
+                buildSchemaName(schemaNamespace, schemaName != null ? schemaName : schema.name()),
```
IIUC, this new `buildSchemaName` method matches the same behavior when a `schema.name` configuration value is provided, but no `schema.namespace` config value is provided: the name is used no matter what the `schema.name` value is. For example, a `schema.name=` config value would use the blank (or empty) name for the schema.

However, if such a user were to simply _add_ a `schema.namespace` config property and keep the `schema.name` set to a blank or empty string, then this method throws a `DataException`.

It might be better to try to keep the behavior as close as possible to the older behavior, except simply use a `schema.namespace` when set to a non-blank or non-empty string. IOW:
```
protected static String buildSchemaName(String namespace, String name) {
    if (Utils.isBlank(namespace)) {
        return name;
    }
    namespace = namespace.trim().replaceAll("[.]$", "");
    return namespace + "." + name;
}
```

pjmagee replied:
When I saw this, I had expected I would be able to set just the `schema.namespace` property and it would prepend this to the existing `schema.name`.

daehokimm replied:
I agree to @pjmagee 's opinion. 

if I want to replace just namespace of multiple schemas, this is not suitable to configure.
this transform is not consider to this situation?

mnegodaev replied:
@daehokimm thanks for your review, I guess [this parameterized test](https://github.com/apache/kafka/pull/11442/files#diff-8470d18c341b6c20f7a6751d2f42af43560525bfeca29bd980ec29d256153205R77) should answer your question.
The case with just namespace set and an empty schema name is acceptable.
Also [this](https://github.com/apache/kafka/pull/11442/files#diff-8470d18c341b6c20f7a6751d2f42af43560525bfeca29bd980ec29d256153205R94) and [this](https://github.com/apache/kafka/pull/11442/files#diff-8470d18c341b6c20f7a6751d2f42af43560525bfeca29bd980ec29d256153205R129) unit-tests should demonstrate this.

rhauch reviewed (commented):
Thanks for the PR, @mnegodaev. I have a few suggestions below.

pjmagee commented:
Would I be correct with the following example too?

```
{
  "transforms" : "TransformSchema",
  "transforms.TransformSchema.type" : "org.apache.kafka.connect.transforms.SetSchemaMetadata$Value",
  "transforms.TransformSchema.schema.namespace" : "added.namespace"
}
```

mnegodaev committed:
Fix config parameter description

Co-authored-by: Randall Hauch <rhauch@gmail.com>

Commit:
Merge remote-tracking branch 'upstream/trunk' into feature/KAFKA-7883

Commit:
KAFKA-7883 make schema namespace prepending tolerant to redundant dots and spaces

make buildSchemaName method tolerant to redundant dots and spaces in schema name and namespace
cover corner cases of the schema update

mnegodaev commented:
> I've marked this as needing a KIP, though I'm not yet sure whether that is the case. These SMTs are technically part of the API, since they appear in STM configurations, which are considered part of the public API. I'll keep looking into this to figure out if this is really the case, but it's probably safer to assume a KIP is needed.

@rhauch thank you for your review, I've really missed the fact I've broken current behavior on empty schema name!
I've fixed this and added more corner tests cases on the schema update. Also I've made an effort to prettify our tests to make them looking more uniform and readable.

pjmagee commented:
@rhauch @mnegodaev  this is looking pretty good now. Would others need to have reviewed this before being merged in? I'm not too clued in on the open source contributions here but it looks like what @mnegodaev has done is excellent work that may solve a lot of issues for people using kafka with avro. :)

mjsax commented:
I am not familiar with Connect code -- @rhauch @kkonstantine @mimaison should be able to help.

pjmagee commented:
@mnegodaev It looks like this is something you would add to this PR a link to a KIP.

mnegodaev commented:
> @mnegodaev It looks like this is something you would add to this PR a link to a KIP.

@pjmagee, I'll try to do may best. Thanks!

Commit:
KAFKA-7883 add missing test-cases with default values

mnegodaev commented:
@rhauch @mimaison @kkonstantine @pjmagee 
If you guys don't mind I'm going to wait till Wednesday and then call a [VOTE] to have the proposal adopted. Thanks!

mnegodaev commented:
@rhauch @mimaison @kkonstantine @pjmagee I've started a vote for the KIP-855: https://lists.apache.org/thread/b0cdofxhl7p84k14wvhtrhd0r1z8qzpk
Please vote for it if you're agree.

pjmagee commented:
@mimaison Hey, It looks like @mnegodaev  has followed all the right procedures and we have the KIP. As someone with more experience on this repo, what are our next steps?

pjmagee commented:
Hi @hachikuji Sorry to pull you in on this PR - I notice you have been active on some merged PR's quite recently. We have had some input from others regarding the process around creating a KIP, which is now completed for this feature, along with an explanation as to why @mnegodaev  has added this feature, along with myself having a usecase for this feature. What would be the next steps in being able to get this merged? It seems like a elegant solution and change to the SMT to support namespaces. Would it be possible for you to maybe include others who may be able to get this merged? I am rather new to this repo so I am unaware of how best to proceed.

mimaison commented:
@pjmagee I commented in the discuss thread last week, see https://lists.apache.org/thread/92y0qqlwb75xqkxnzwbw7jywfdw8gs8k. Can you reply to my question there?

mnegodaev commented:
> @pjmagee I commented in the discuss thread last week, see https://lists.apache.org/thread/92y0qqlwb75xqkxnzwbw7jywfdw8gs8k. Can you reply to my question there?

I've fixed this typo into KIP, thanks for finding this.

benru89 commented:
whats the status of this?

mimaison commented:
To be able to merge this, the associated [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-855%3A+Add+schema.namespace+parameter+to+SetSchemaMetadata+SMT+in+Kafka+Connect) must be voted.
In the [discussion thread](https://lists.apache.org/thread/3hkd9lljobf9rl56ogjpcbo4ldoxcz5n) for the KIP, a few questions have not been answered yet by the author. Until these are answered, it's unlikely there will be any votes, so this will stay stuck.

mnegodaev commented:
Hi guys, I'll try to answer the questions in the near time, thanks.

ср, 15 февр. 2023 г. в 15:51, Mickael Maison ***@***.***>:

> To be able to merge this, the associated KIP
> <https://cwiki.apache.org/confluence/display/KAFKA/KIP-855%3A+Add+schema.namespace+parameter+to+SetSchemaMetadata+SMT+in+Kafka+Connect>
> must be voted.
> In the discussion thread
> <https://lists.apache.org/thread/3hkd9lljobf9rl56ogjpcbo4ldoxcz5n> for
> the KIP, a few questions have not been answered yet by the author. Until
> these are answered, it's unlikely there will be any votes, so this will
> stay stuck.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/kafka/pull/11442#issuecomment-1431141433>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2RTBY2JST3W745OFVSLJ3WXSYLTANCNFSM5G3C2JEA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

MartinJohannessen commented:
Guys, please. I need this. I'm having to literally make a hundred individual worker configurations instead of one, because I don't have this.

mimaison commented:
@MartinJohannessen This is an open source project. If an issue or feature matters to you, you can participate and help.

In this case, @mnegodaev has done most of the work already but the KIP has to be approved by at least 3 committers to be accepted. At the moment, I'm the only committer to have voted for it. As said in [my comment above](https://github.com/apache/kafka/pull/11442#issuecomment-1431141433) there are a bunch of questions asked by another committer that were never answered, hence probably why he, and others, did not vote.

Until these questions are answered the proposal is stuck. I'm not sure if @mnegodaev plans to ever return to this or not. Nobody else from the community expressed interest in the [DISCUSS](https://lists.apache.org/thread/3hkd9lljobf9rl56ogjpcbo4ldoxcz5n) thread. You can take this over and try to complete it.
